pkg/ring/model.go

```diff
@@ -184,7 +184,7 @@ func (d *Desc) TokensFor(id string) (tokens, other []uint32) {
 func (i *IngesterDesc) IsHealthy(op Operation, heartbeatTimeout time.Duration) bool {
 	if op == Write && i.State != ACTIVE {
 		return false
-	} else if op == Read && i.State == JOINING {
+	} else if op == Read && i.State != ACTIVE && i.State != LEAVING {
```

Comment:
IIRC pending ingesters shouldn't have any tokens; how did you trigger this?

Reply:
It should occur in the paths where we use `forAllIngesters()`. I think one path is:
1. Querier receive a `LabelNames` request
2. Querier forwards the request to the distributor
3. The distributor runs `client.LabelNames()` in `distributor.forAllIngesters()`
4. `distributor.forAllIngesters()` calls `ring.GetAll()`
5. `ring.GetAll()` returns all ingesters which satisfy the `IsHealthy()`

Reply:
@bboreham What's your take on this? _If this change is a problem, I could rollback it in order to move one with this PR._

Reply:
Thanks for enumerating these.